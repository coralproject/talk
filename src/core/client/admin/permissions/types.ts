import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
