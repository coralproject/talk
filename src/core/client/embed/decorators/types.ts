import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
