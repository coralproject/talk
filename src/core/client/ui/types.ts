import React from "react";

// TODO: Extract useful common types into its own package.

/**
 * Overwrite properties of `T`.
 *
 * E.g. Omit<{a: boolean, b: boolean}, "b"> = {a: boolean}
 */
export type Omit<U, K extends keyof U> = Pick<U, Exclude<keyof U, K>>;

/**
 * Overwrite properties of `T`.
 *
 * E.g. Overwrite<{a: boolean}, {a: string}> = {a: string}
 */
export type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> & U;

/**
 * Returns the PropTypes from a React Component.
 *
 * E.g. type ButtonProps = ReturnPropTypes<Button>;
 */
export type PropTypesOf<T> = T extends React.ComponentType<infer R>
  ? R
  : T extends React.Component<infer S> ? S : {};
