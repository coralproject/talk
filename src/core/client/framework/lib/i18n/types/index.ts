export { default as FluentShortNumber } from "./FluentShortNumber";
