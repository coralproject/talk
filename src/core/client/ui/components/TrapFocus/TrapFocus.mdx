---
name: TrapFocus
menu: UI Kit
---

import { Playground } from 'docz'
import TrapFocus from './TrapFocus'
import Button from '../Button'

# TrapFocus

Traps focus inside component when using keyboard navigation for accessibility purposes.

## Basic usage

<Playground>
  <div>
    <TrapFocus>
    {({firstFocusableRef, lastFocusableRef}) =>
      <div>
        <Button forwardRef={firstFocusableRef} autoFocus>Cancel</Button>
        <Button forwardRef={lastFocusableRef}>Send</Button>
      </div>
    }
    </TrapFocus>
  </div>
</Playground>
