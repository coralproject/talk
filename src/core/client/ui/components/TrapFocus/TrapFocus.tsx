import React, { RefObject } from "react";

export interface Focusable {
  focus: () => void;
}

interface RenderProps {
  firstFocusableRef: RefObject<Focusable>;
  lastFocusableRef: RefObject<Focusable>;
}

export interface TrapFocusProps {
  children: (props: RenderProps) => React.ReactNode;
}

export default class TrapFocus extends React.Component<TrapFocusProps> {
  private firstFocusableRef = React.createRef<Focusable>();
  private lastFocusableRef = React.createRef<Focusable>();

  // Trap keyboard focus inside the dropdown until a value has been chosen.
  private focusBegin = () => this.firstFocusableRef.current!.focus();
  private focusEnd = () => this.lastFocusableRef.current!.focus();

  public render() {
    return (
      <>
        <div tabIndex={0} onFocus={this.focusEnd} />
        {this.props.children({
          firstFocusableRef: this.firstFocusableRef,
          lastFocusableRef: this.lastFocusableRef,
        })}
        <div tabIndex={0} onFocus={this.focusBegin} />
      </>
    );
  }
}
