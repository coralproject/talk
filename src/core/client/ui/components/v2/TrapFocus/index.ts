export { default, TrapFocusProps } from "./TrapFocus";
