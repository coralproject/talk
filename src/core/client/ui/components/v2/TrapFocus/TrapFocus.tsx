/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
import React, { FunctionComponent, RefObject, useEffect, useRef } from "react";

import { useUIContext } from "coral-ui/components/v2/UIContext";

interface RenderProps {
  firstFocusableRef: RefObject<any>;
  lastFocusableRef: RefObject<any>;
}

type RenderPropsCallback = (props: RenderProps) => React.ReactNode;

interface TrapFocusProps {
  children?: React.ReactNode | RenderPropsCallback;
}

function isRenderProp(
  children: TrapFocusProps["children"]
): children is RenderPropsCallback {
  return typeof children === "function";
}

const TrapFocus: FunctionComponent<TrapFocusProps> = ({ children }) => {
  const { renderWindow } = useUIContext();
  const fallbackRef = useRef<HTMLDivElement>(null);
  const firstFocusableRef = useRef<HTMLElement>(null);
  const lastFocusableRef = useRef<HTMLElement>(null);
  const previousActiveElement: any | null = renderWindow.document.activeElement;

  // Trap keyboard focus inside the dropdown until a value has been chosen.
  const focusBegin = () => {
    if (firstFocusableRef.current) {
      firstFocusableRef.current.focus();
    } else {
      if (fallbackRef.current) {
        fallbackRef.current.focus();
      }
    }
  };

  const focusEnd = () => {
    if (lastFocusableRef.current) {
      lastFocusableRef.current.focus();
    } else {
      if (fallbackRef.current) {
        fallbackRef.current.focus();
      }
    }
  };

  useEffect(() => {
    if (fallbackRef.current) {
      fallbackRef.current.focus();
    }
    return () => {
      if (previousActiveElement && previousActiveElement.focus) {
        previousActiveElement.focus();
      }
    };
  }, []);

  return (
    <>
      <div tabIndex={0} onFocus={focusEnd} />
      <div tabIndex={-1} ref={fallbackRef} />
      {isRenderProp(children)
        ? children({
            firstFocusableRef,
            lastFocusableRef,
          })
        : children}
      <div tabIndex={0} onFocus={focusBegin} />
    </>
  );
};

export default TrapFocus;
