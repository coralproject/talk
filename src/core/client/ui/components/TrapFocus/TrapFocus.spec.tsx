import React from "react";
import { create } from "react-test-renderer";
import Sinon from "sinon";

import { PropTypesOf } from "talk-ui/types";
import TrapFocus from "./TrapFocus";

const FakeFocusable: any = class extends React.Component {
  public focus = Sinon.spy();
  public render() {
    return null;
  }
};

it("renders correctly", () => {
  const props: PropTypesOf<TrapFocus> = {
    children: () => (
      <>
        <span>child1</span>
        <span>child2</span>
      </>
    ),
  };
  const renderer = create(
    <div>
      <TrapFocus {...props} />
    </div>
  );
  expect(renderer.toJSON()).toMatchSnapshot();
});

it("Change focus to `lastFocusable` when focus reaches beginning", () => {
  const props: PropTypesOf<TrapFocus> = {
    children: ({ firstFocusableRef, lastFocusableRef }) => (
      <>
        <FakeFocusable ref={firstFocusableRef} />
        <span>child1</span>
        <span>child2</span>
        <FakeFocusable ref={lastFocusableRef} />
      </>
    ),
  };
  const renderer = create(
    <div>
      <TrapFocus {...props} />
    </div>
  );
  renderer.root.findAllByProps({ tabIndex: 0 })[0].props.onFocus();
  expect(
    renderer.root.findAllByType(FakeFocusable)[0].instance.focus.called
  ).toBe(false);
  expect(
    renderer.root.findAllByType(FakeFocusable)[1].instance.focus.called
  ).toBe(true);
});

it("Change focus to `firstFocusable` when focus reaches end", () => {
  const props: PropTypesOf<TrapFocus> = {
    children: ({ firstFocusableRef, lastFocusableRef }) => (
      <>
        <FakeFocusable ref={firstFocusableRef} />
        <span>child1</span>
        <span>child2</span>
        <FakeFocusable ref={lastFocusableRef} />
      </>
    ),
  };
  const renderer = create(
    <div>
      <TrapFocus {...props} />
    </div>
  );
  renderer.root.findAllByProps({ tabIndex: 0 })[1].props.onFocus();
  expect(
    renderer.root.findAllByType(FakeFocusable)[0].instance.focus.called
  ).toBe(true);
  expect(
    renderer.root.findAllByType(FakeFocusable)[1].instance.focus.called
  ).toBe(false);
});
