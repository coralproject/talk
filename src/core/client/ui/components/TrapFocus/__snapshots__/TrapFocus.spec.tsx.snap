// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    onFocus={[Function]}
    tabIndex={0}
  />
  <span>
    child1
  </span>
  <span>
    child2
  </span>
  <div
    onFocus={[Function]}
    tabIndex={0}
  />
</div>
`;
